dynet/training.cc

```diff
@@ -279,7 +370,213 @@ void Trainer::populate(std::istream& is, real lr)
+{
+    tvec(*ema).device(*dev.edevice) = ema_beta * tvec(*ema) + (1-ema_beta) * tvec(*p);
+}
+
+template<class MyDevice>
+void Trainer::swap_params_to_ema_rule_dev(
+    const MyDevice& dev,
+    bool bias_correction, bool save_weights,
+    Tensor* p, Tensor* mem, Tensor* ema
+)
+{
+    if (save_weights)
+        tvec(*mem).device(*dev.edevice) = tvec(*p);
+
+    if (bias_correction)
+    {
+        const real pow_beta = pow(ema_beta, ema_updates);
+        const real scale = 1.f / (1.f - pow_beta);
+        tvec(*p).device(*dev.edevice) = scale * tvec(*ema);
+    }
+    else
+        tvec(*p).device(*dev.edevice) = tvec(*ema);
+}
+
+template<class MyDevice>
+void Trainer::swap_params_to_weights_rule_dev(const MyDevice& dev, Tensor* p, Tensor* mem)
+{
+    tvec(*p).device(*dev.edevice) = tvec(*mem);
+}
+
+#ifndef __CUDACC__
+
+bool Trainer::ema()
+{
+    return ema_beta > 0.f;
+}
+
+void Trainer::ema(float beta, unsigned update_freq)
+{
+    if (updates > 0)
+        DYNET_RUNTIME_ERR("This function must be called before any update");
+    if (update_freq == 0u)
+        DYNET_RUNTIME_ERR("The update frequency cannot be null");
+
+    ema_beta = beta;
+    ema_update_freq = update_freq;
+}
+
+void Trainer::swap_params_to_ema(bool bias_correction, bool save_weights)
+{
+    if (!ema())
+        DYNET_RUNTIME_ERR("EMA is not enabled");
+    if (ema_updates == 0u)
+        DYNET_RUNTIME_ERR("EMA has not been set yet");
+
+    if (ema_params_swapped)
+        return; // nothing to do
+    ema_params_swapped = true;
+    ema_params_saved = save_weights;
+
+    const auto& params = model->parameters_list();
+    const auto& lparams = model->lookup_parameters_list();
+
+    // check memory (shadow params are automatically initialized to zero)
+    if (ema_p.size() < params.size())
+        allocate_shadow_parameters(*model, ema_p.size(), ema_p);
+    if (ema_lp.size() < lparams.size())
+        allocate_shadow_lookup_parameters(*model, ema_lp.size(), ema_lp);
+
+    if (save_weights)
+    {
+        if (ema_saved_p.size() < params.size())
+            allocate_shadow_parameters(*model, ema_saved_p.size(), ema_saved_p);
+        if (ema_saved_lp.size() < lparams.size())
+            allocate_shadow_lookup_parameters(*model, ema_saved_lp.size(), ema_saved_lp);
+    }
+
+    for(size_t i = 0; i < params.size(); ++i)
+    {
+        Tensor& weights = params[i]->values;
+        Tensor& mem = ema_saved_p[i].h;
+        Tensor& ema = ema_p[i].h;
+
+        swap_params_to_ema_rule(bias_correction, save_weights, &weights, &mem, &ema);
+    }
+    for(size_t i = 0; i < lparams.size(); ++i)
+    {
+        Tensor& weights = lparams[i]->all_values;
+        Tensor& mem = ema_saved_lp[i].all_h;
+        Tensor& ema = ema_lp[i].all_h;
+
+        swap_params_to_ema_rule(bias_correction, save_weights, &weights, &mem, &ema);
+    }
+}
 
+void Trainer::swap_params_to_weights()
+{
+    if (!ema_params_swapped)
+        return;
```

Comment:
@FilippoC Aside from the compilation issues above, I think the reason for why you aren't getting results is the lack of `ema_params_swapped = false;` here to undo the `ema_params_swapped = true;` on line 515. 

I made this fix and the code seems to working perfectly! :) 

Thanks a bunch for this. Using CMA as I wished was absolutely trivial. I simply changed the update rule to `tvec(*ema).device(*dev.edevice) = (real(ema_updates) * tvec(*ema) + tvec(*p)) / (real(ema_updates)+1);`.